x/bsonx/bsoncore/bsoncore.go

```diff
@@ -729,8 +732,11 @@ func readi32(src []byte) (int32, []byte, bool) {
 	if len(src) < 4 {
 		return 0, src, false
 	}
-
-	return (int32(src[0]) | int32(src[1])<<8 | int32(src[2])<<16 | int32(src[3])<<24), src[4:], true
+	ln := (int32(src[0]) | int32(src[1])<<8 | int32(src[2])<<16 | int32(src[3])<<24)
+	if ln < 1 {
```

Comment:
@tvi This `ln < 1` check is causing our tests to fail because it doesn't allow a number below 1 to appear as a BSON int32, but this is not a valid constraint. BSON int32s are signed and can be 0 or negative. Specifically, all of our integration tests fail with this PR because the server reports `minWireVersion: 0` in it's `isMaster` responses.

Can you explain the motivation for this constraint?

Reply:
I do not have ability to see your integration tests, but feel free to reproduce with test cases I added. Currently the library will panic with such an int.

Reply:
@tvi We think we've figured out the issue here. `readi32` is a general-purpose function to read 32-bit ints, which can be 0 or negative. We think this check belongs in `ReadLength` because it makes sense to constrain lengths to >= 1.